### PR TITLE
Optimize closure iterator locals

### PR DESCRIPTION
**Pull request by yglukhov**
This pr redefines the relation between lambda lifting and closureiter transformation.

Key takeaways:
- Lambdalifting now has less distinction between closureiters and regular closures. Namely instead of lifting _all_ closureiter variables, it lifts only those variables it would also lift for simple closure, i.e. those not owned by the closure.
- It is now closureiter transformation's responsibility to lift all the locals that need lifting and are not lifted by lambdalifting. So now we lift only those locals that appear in more than one state. The rest remains on stack, yay!
- Closureiter transformation always relies on the closure env param created by lambdalifting. Special care taken to make lambdalifting create it even in cases when it's "too early" to lift.
- Environments created by lambdalifting will contain `:state` only for closureiters, whereas previously any closure env contained it.

IMO this is a more reasonable approach as it simplifies not only lambdalifting, but transf too (e.g. freshVarsForClosureIters is now gone for good).

I tried to organize the changes logically by commits, so it might be easier to review this on per commit basis.

Some ugliness:
- Adding lifting to closureiters transformation I had to repeat this matching of `return result = value` node. I tried to understand why it is needed, but that was just another rabbit hole, so I left it for another time. @Araq your input is welcome.
- In the last commit I've reused currently undocumented `liftLocals` pragma for symbols so that closureiter transformation will forcefully lift those even if they don't require lifting otherwise. This is needed for [yasync](https://github.com/yglukhov/yasync) or else it will be very sad.

Overall I'm quite happy with the results, I'm seeing some noticeable code size reductions in my projects. Heavy closureiter/async users, please give it a go.

### CODE REVIEW

File: compiler/closureiters.nim [hunk 1]
```diff
@@ -255,9 +240,18 @@ proc addGotoOut(n: PNode, gotoOut: PNode): PNode =
...
+  var v = initialValue
+  if v == nil:
+    v = ctx.g.emptyNode
+  newTree(nkVarSection, newTree(nkIdentDefs, newSymNode(s), ctx.g.emptyNode, v))
+
+proc newTempVar(ctx: var Ctx, typ: PType, parent: PNode, initialValue: PNode = nil): PSym =
+  result = newSym(skVar, getIdent(ctx.g.cache, ":tmpSlLower" & $ctx.tempVarId), ctx.idgen, ctx.fn, ctx.fn.info)
   inc ctx.tempVarId
+  result.typ = typ
+  assert(not typ.isNil)
```
Comment (juancarlospaco) [inline compiler/closureiters.nim]: If someday this assert does trigger, would be better to have a message about what happened, plz add a message to the assert IMHO.

**Comment by mratsim** [comment]
oooooohh so for closure iterators used as "functional programming emulator" and futures that don't escape their scope there is no heap allocation?

**Comment by yglukhov** [comment]
@mratsim no, in terms of allocation nothing changes. It's only the allocated environments will be smaller in the following case:
```nim
iterator foo() {.closure.} =
  var a = 1 # Previously was part of env. Now it is on stack
  var b = 2 # Previously was part of env, and remains there because it lives through yield
  echo a
  yield
  echo b
```
 What you're asking for sounds like what [yasync](https://github.com/yglukhov/yasync) is doing?

**Comment by SirOlaf** [comment]
I haven't done any performance tests, but it works for what I was trying to fix in my pr. Very nice changes.
```nim
type Foo = object

proc `=copy`(a: var Foo, b: Foo) {.error.}

proc consumer(x: sink Foo) = discard

iterator iter(): int {.closure.} =
  var foo = Foo()
  consumer(foo)

for x in iter():
  discard
```

**Comment by arnetheduck** [comment]
ooh, nice to see - have you run this through the chronos test suite?

**Comment by yglukhov** [comment]
@arnetheduck I tried, but getting errors even with nim devel:
```
tlsstream.nim(515, 43) Error: undeclared identifier: 'X509ClassPointerConst'
```
Please do me a favor and test and report. If you find any issues I'll have no choice but to look into it)

**Comment by ringabout** [comment]
> have you run this through the chronos test suite?

> pkg "chronos", "nim c -r -d:release tests/testall"

Nim packages CI has covered chronos testing suit with ORC

File: compiler/pragmas.nim [hunk 2]
```diff
@@ -1308,7 +1308,8 @@ proc singlePragma(c: PContext, sym: PSym, n: PNode, i: var int,
         noVal(c, it)
         if sym == nil: invalidPragma(c, it)
         else: sym.flags.incl sfUsed
-      of wLiftLocals: discard
+      of wLiftLocals:
+        sym.flags.incl(sfForceLift)
```
Comment (Araq) [inline compiler/pragmas.nim:1312]: Why is this required?

Comment (yglukhov) [reply compiler/pragmas.nim:1312]: The yasync library injects some variables into the closureiter body and expects they will be available in its env. After this pr this is no longer guaranteed. Hence I've added an escape hatch in form of `{.liftLocals.}` pragma for yasync to take advantage of it. I hope to find a better way in the future, but that likely will happen after state-split algorithm changes. As a side note you might be better off by reviewing just the commit "[liftLocals pragma for closureiter locals to be forcefully lifted" to better see the affected parts.

Comment (Araq) [reply compiler/pragmas.nim:1312]: Well it's not documented and if your library needs this annotation it implies bad things:

1. This is not 100% backwards compatible. (However, the change is beneficial enough to warrant it.)
2. This is not a safe transformation.

What does yasync do that requires the new annotation?

Comment (juancarlospaco) [reply compiler/pragmas.nim:1312]: is this new pragma documented?

**Comment by arnetheduck** [comment]
> @arnetheduck I tried, but getting errors even with nim devel:
> 
> ```
> tlsstream.nim(515, 43) Error: undeclared identifier: 'X509ClassPointerConst'
> ```

this is an out-of-date bearssl in your global nimble folder - needs wiping, same as it's been for the past 6 years ;)

File: compiler/pragmas.nim (hunk 2 above)
Comment (yglukhov) [reply compiler/pragmas.nim:1312]: It has never been documented in the first place. `liftLocals` was invented to facilitate some experiments in the library/async/cps space, which are now abandoned and it is pretty much unused by anyone, correct me if I'm wrong. So I just gave this pragma another breath to be useful in yasync, I really don't expect anyone else to use it, just like it has been this way up to this point :)

> This is not 100% backwards compatible

Why not? It is backwards-compatible because previously it was allowed only for functions (and the old behavior remains). Now it is also allowed for variables with slightly different-but-close-enough meaning. If you're concerned that it will be removed or changed in the future, I don't mind, we'll figure it out when time comes.

> This is not a safe transformation.

Why not? You can place `{.liftLocals.}` willy-nilly on any variable and worst harm it will do is lift said local to the env, just like every other local was lifted until this pr.

> What does yasync do that requires the new annotation?

I don't know how detailed answer you're expecting, but yasync implements async and Futures by collapsing all the futures and closureiter environments in the callstack into one big environment which allows running async tasks with a single allocation per whole task, which can also happen on the stack in certain scenarios. For that to happen yasync injects the Continuation header, result storage and arguments into the closureiter code as variables, and later expects to find them lifted in the corresponding environment.

**Comment by yglukhov** [comment]
@arnetheduck ok that worked and it passes.
